`patrol test` not using Java version reported by `flutter doctor`

### Steps to reproduce

On a Mac, with latest AndroidStudio having JBR 17 as part of the AndroidStudio instalation

Running: `patrol test --flavor dev --verbose`

Fails with Unsupported class file major version 65.

I looked at the version codes and 65 is produced by Java 21 which is not part of Android Studio.

I followed the native setup and got it working fine for iPhones, but Android does not work.

Building the app with `flutter build apk --flavor dev` works fine.

### Actual results

The test app fails to compile

### Logs

<details>
<summary>Logs</summary>

```
patrol test --flavor dev --verbose
Verbose mode enabled. More logs will be printed.
$ flutter --suppress-analytics --no-version-check pub deps --style=list
Received 1 test target(s)
Received test target: /Users/michalhazdra/Work/ACO/aco_dsa_mobile/integration_test/login/login_test.dart
Generated entrypoint /Users/michalhazdra/Work/ACO/aco_dsa_mobile/integration_test/test_bundle.dart with 1 bundled test(s)
Received Android flavor: dev
Received iOS flavor: dev
Received macOS flavor: dev
$ flutter --no-version-check --suppress-analytics devices --machine
No device specified, using the first one (LMG810EAWbedd125a)
Received 1 device(s) to run on
Received device: LMG810EAWbedd125a
Received 9 --dart-define(s) (0 custom, 9 internal)
Received internal --dart-define: PATROL_WAIT=0
Received internal --dart-define: PATROL_APP_PACKAGE_NAME=cz.aco.dsa_mobile
Received internal --dart-define: PATROL_APP_BUNDLE_ID=aco.dev
Received internal --dart-define: PATROL_ANDROID_APP_NAME=ACO DSA DEV
Received internal --dart-define: PATROL_IOS_APP_NAME=ACO DSA DEV
Received internal --dart-define: INTEGRATION_TEST_SHOULD_REPORT_RESULTS_TO_NATIVE=false
Received internal --dart-define: PATROL_TEST_LABEL_ENABLED=true
Received internal --dart-define: PATROL_TEST_SERVER_PORT=8081
Received internal --dart-define: PATROL_APP_SERVER_PORT=8082
• Building apk with entrypoint test_bundle.dart...
$ ./gradlew :app:assembleDevDebug -Ptarget=/Users/michalhazdra/Work/ACO/aco_dsa_mobile/integration_test/test_bundle.dart -Pdart-defines=UEFUUk9MX1dBSVQ9MA==,UEFUUk9MX0FQUF9QQUNLQUdFX05BTUU9Y3ouYWNvLmRzYV9tb2JpbGU=,UEFUUk9MX0FQUF9CVU5ETEVfSUQ9YWNvLmRldg==,UEFUUk9MX0FORFJPSURfQVBQX05BTUU9QUNPIERTQSBERVY=,UEFUUk9MX0lPU19BUFBfTkFNRT1BQ08gRFNBIERFVg==,SU5URUdSQVRJT05fVEVTVF9TSE9VTERfUkVQT1JUX1JFU1VMVFNfVE9fTkFUSVZFPWZhbHNl,UEFUUk9MX1RFU1RfTEFCRUxfRU5BQkxFRD10cnVl,UEFUUk9MX1RFU1RfU0VSVkVSX1BPUlQ9ODA4MQ==,UEFUUk9MX0FQUF9TRVJWRVJfUE9SVD04MDgy -Papp-server-port=8082 -Ptest-server-port=8081
        : > Task :gradle:compileJava NO-SOURCE
        : > Task :gradle:compileGroovy FAILED

        FAILURE: Build failed with an exception.

        * What went wrong:
        Execution failed for task ':gradle:compileGroovy'.
        > BUG! exception in phase 'semantic analysis' in source unit '/Users/michalhazdra/Flutter/flutter/packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy' Unsupported class file major version 65

        * Try:
        > Run with --stacktrace option to get the stack trace.
        > Run with --info or --debug option to get more log output.
        > Run with --scan to get full insights.

        * Get more help at https://help.gradle.org

        BUILD FAILED in 3s
        : 1 actionable task: 1 executed
✗ Failed to build apk with entrypoint test_bundle.dart (Gradle build failed with code 1) (3.7s)
Exception: Gradle build failed with code 1
#0      AndroidTestBackend.build.<anonymous closure> (package:patrol_cli/src/android/android_test_backend.dart:67:9)
<asynchronous suspension>
#1      DisposeScope.run (package:dispose_scope/src/dispose_scope.dart:46:7)
<asynchronous suspension>
#2      AndroidTestBackend.build (package:patrol_cli/src/android/android_test_backend.dart:42:5)
<asynchronous suspension>
#3      TestCommand._build (package:patrol_cli/src/commands/test.dart:268:7)
<asynchronous suspension>
#4      TestCommand.run (package:patrol_cli/src/commands/test.dart:202:5)
<asynchronous suspension>
#5      CommandRunner.runCommand (package:args/command_runner.dart:212:13)
<asynchronous suspension>
#6      PatrolCommandRunner.runCommand (package:patrol_cli/src/runner/patrol_command_runner.dart:341:18)
<asynchronous suspension>
#7      PatrolCommandRunner.run (package:patrol_cli/src/runner/patrol_command_runner.dart:285:18)
<asynchronous suspension>
#8      patrolCommandRunner (package:patrol_cli/src/runner/patrol_command_runner.dart:70:20)
<asynchronous suspension>
#9      main (file:///Users/michalhazdra/.pub-cache/hosted/pub.dev/patrol_cli-2.6.5/bin/main.dart:6:20)
<asynchronous suspension>

See the logs above to learn what happened. Also consider running with --verbose. If the logs still aren't useful, then it's a bug - please report it.
Exception: Gradle build failed with code 1
#0      AndroidTestBackend.build.<anonymous closure> (package:patrol_cli/src/android/android_test_backend.dart:67:9)
<asynchronous suspension>
#1      DisposeScope.run (package:dispose_scope/src/dispose_scope.dart:46:7)
<asynchronous suspension>
#2      AndroidTestBackend.build (package:patrol_cli/src/android/android_test_backend.dart:42:5)
<asynchronous suspension>
#3      TestCommand._build (package:patrol_cli/src/commands/test.dart:268:7)
<asynchronous suspension>
#4      TestCommand.run (package:patrol_cli/src/commands/test.dart:202:5)
<asynchronous suspension>
#5      CommandRunner.runCommand (package:args/command_runner.dart:212:13)
<asynchronous suspension>
#6      PatrolCommandRunner.runCommand (package:patrol_cli/src/runner/patrol_command_runner.dart:341:18)
<asynchronous suspension>
#7      PatrolCommandRunner.run (package:patrol_cli/src/runner/patrol_command_runner.dart:285:18)
<asynchronous suspension>
#8      patrolCommandRunner (package:patrol_cli/src/runner/patrol_command_runner.dart:70:20)
<asynchronous suspension>
#9      main (file:///Users/michalhazdra/.pub-cache/hosted/pub.dev/patrol_cli-2.6.5/bin/main.dart:6:20)
<asynchronous suspension>
```

</details>


### Patrol version

patrol: ^3.6.0

### Patrol Doctor output

<details>
<summary>Patrol Doctor output</summary>

```
Patrol doctor:
Patrol CLI version: 2.6.5
Android: 
• Program adb found in /opt/local/bin/adb
• Env var $ANDROID_HOME set to /Users/michalhazdra/Library/Android/sdk
iOS / macOS: 
• Program xcodebuild found in /usr/bin/xcodebuild
• Program ideviceinstaller found in /opt/homebrew/bin/ideviceinstaller
```

</details>


### Flutter Doctor output

<details>
<summary>Flutter Doctor output</summary>

```
[✓] Flutter (Channel stable, 3.19.2, on macOS 14.3 23D56 darwin-arm64, locale en-CZ)
    • Flutter version 3.19.2 on channel stable at /Users/michalhazdra/Flutter/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7482962148 (4 weeks ago), 2024-02-27 16:51:22 -0500
    • Engine revision 04817c99c9
    • Dart version 3.3.0
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/michalhazdra/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/michalhazdra/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C500b
    • CocoaPods version 1.14.3

[✗] Chrome - develop for the web (Cannot find Chrome executable at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Android Studio (version 2023.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.3)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.87.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.84.0

[✓] Connected device (2 available)

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>

--- a/packages/patrol_cli/CHANGELOG.md
+++ b/packages/patrol_cli/CHANGELOG.md
@@ -1,3 +1,7 @@
+## 2.9.0
+
+- Use `java` version from `flutter doctor`
+
 ## 2.8.1
 
 - Fix parsing non string values from --dart-define-from-file (#2243).

--- a/packages/patrol_cli/CHANGELOG.md
+++ b/packages/patrol_cli/CHANGELOG.md
@@ -1,6 +1,6 @@
-## 2.9.0
+## 3.0.0
 
-- Use `java` version from `flutter doctor`
+- **Breaking:** Use `java` version from `flutter doctor`
 
 ## 2.8.1
 

--- a/packages/patrol_cli/lib/src/android/android_test_backend.dart
+++ b/packages/patrol_cli/lib/src/android/android_test_backend.dart
@@ -1,3 +1,4 @@
+import 'dart:async';
 import 'dart:io' show Process;
 
 import 'package:adb/adb.dart';
@@ -37,8 +38,11 @@ class AndroidTestBackend {
   final FileSystem _fs;
   final DisposeScope _disposeScope;
   final Logger _logger;
+  late final String javaPath;
 
   Future<void> build(AndroidAppOptions options) async {
+    await loadJavaPathFromFlutterDoctor();
+
     await _disposeScope.run((scope) async {
       final subject = options.description;
       final task = _logger.task('Building $subject');
@@ -52,6 +56,9 @@ class AndroidTestBackend {
         options.toGradleAssembleInvocation(isWindows: _platform.isWindows),
         runInShell: true,
         workingDirectory: _fs.currentDirectory.childDirectory('android').path,
+        environment: {
+          'JAVA_HOME': javaPath,
+        },
       )
         ..disposedBy(scope);
       process.listenStdOut((l) => _logger.detail('\t: $l')).disposedBy(scope);
@@ -73,6 +80,7 @@ class AndroidTestBackend {
         options.toGradleAssembleTestInvocation(isWindows: _platform.isWindows),
         runInShell: true,
         workingDirectory: _fs.currentDirectory.childDirectory('android').path,
+        environment: {'JAVA_HOME': javaPath},
       )
         ..disposedBy(scope);
       process.listenStdOut((l) => _logger.detail('\t: $l')).disposedBy(scope);
@@ -90,6 +98,33 @@ class AndroidTestBackend {
         throw Exception(cause);
       }
     });
+  }
+
+  Future<void> loadJavaPathFromFlutterDoctor() async {
+    final javaCompleterPath = Completer<String?>();
+
+    await _disposeScope.run((scope) async {
+      final process = await _processManager.start(
+        ['flutter', 'doctor', '--verbose'],
+        runInShell: true,
+      )
+        ..disposedBy(scope);
+
+      process.listenStdOut(
+        (line) async {
+          if (line.contains('• Java binary at:') &&
+              javaCompleterPath.isCompleted == false) {
+            final path = line
+                .replaceAll('• Java binary at:', '')
+                .replaceAll('/bin/java', '')
+                .trim();
+            javaCompleterPath.complete(path);
+          }
+        },
+      ).disposedBy(scope);
+    });
+
+    javaPath = await javaCompleterPath.future ?? '';
   }
 
   /// Executes the tests of the given [options] on the given [device].
@@ -110,7 +145,10 @@ class AndroidTestBackend {
       final process = await _processManager.start(
         options.toGradleConnectedTestInvocation(isWindows: _platform.isWindows),
         runInShell: true,
-        environment: {'ANDROID_SERIAL': device.id},
+        environment: {
+          'ANDROID_SERIAL': device.id,
+          'JAVA_HOME': javaPath,
+        },
         workingDirectory: _fs.currentDirectory.childDirectory('android').path,
       )
         ..disposedBy(scope);

--- a/packages/patrol_cli/lib/src/android/android_test_backend.dart
+++ b/packages/patrol_cli/lib/src/android/android_test_backend.dart
@@ -41,7 +41,7 @@ class AndroidTestBackend {
   late final String javaPath;
 
   Future<void> build(AndroidAppOptions options) async {
-    await loadJavaPathFromFlutterDoctor();
+    await loadJavaPathFromFlutterDoctor(options.flutter.command.executable);
 
     await _disposeScope.run((scope) async {
       final subject = options.description;
@@ -80,7 +80,9 @@ class AndroidTestBackend {
         options.toGradleAssembleTestInvocation(isWindows: _platform.isWindows),
         runInShell: true,
         workingDirectory: _fs.currentDirectory.childDirectory('android').path,
-        environment: {'JAVA_HOME': javaPath},
+        environment: {
+          'JAVA_HOME': javaPath,
+        },
       )
         ..disposedBy(scope);
       process.listenStdOut((l) => _logger.detail('\t: $l')).disposedBy(scope);
@@ -100,12 +102,16 @@ class AndroidTestBackend {
     });
   }
 
-  Future<void> loadJavaPathFromFlutterDoctor() async {
+  Future<void> loadJavaPathFromFlutterDoctor(String commandExecutable) async {
     final javaCompleterPath = Completer<String?>();
 
     await _disposeScope.run((scope) async {
       final process = await _processManager.start(
-        ['flutter', 'doctor', '--verbose'],
+        [
+          commandExecutable,
+          'doctor',
+          '--verbose',
+        ],
         runInShell: true,
       )
         ..disposedBy(scope);
@@ -116,7 +122,10 @@ class AndroidTestBackend {
               javaCompleterPath.isCompleted == false) {
             final path = line
                 .replaceAll('• Java binary at:', '')
-                .replaceAll('/bin/java', '')
+                .replaceAll(
+                  _platform.isWindows ? r'\bin\java' : '/bin/java',
+                  '',
+                )
                 .trim();
             javaCompleterPath.complete(path);
           }

--- a/packages/patrol_cli/lib/src/base/constants.dart
+++ b/packages/patrol_cli/lib/src/base/constants.dart
@@ -1,3 +1,3 @@
 /// Version of Patrol CLI. Must be kept in sync with pubspec.yaml.
 /// If you update this, make sure that compatibility-table.mdx is updated (if needed)
-const version = '2.8.1';
+const version = '2.9.0';

--- a/packages/patrol_cli/lib/src/base/constants.dart
+++ b/packages/patrol_cli/lib/src/base/constants.dart
@@ -1,3 +1,3 @@
 /// Version of Patrol CLI. Must be kept in sync with pubspec.yaml.
 /// If you update this, make sure that compatibility-table.mdx is updated (if needed)
-const version = '2.9.0';
+const version = '3.0.0';

--- a/packages/patrol_cli/lib/src/commands/develop.dart
+++ b/packages/patrol_cli/lib/src/commands/develop.dart
@@ -94,10 +94,6 @@ class DevelopCommand extends PatrolCommand {
       ),
     );
 
-    await _compatibilityChecker.checkVersionsCompatibility(
-      flutterCommand: flutterCommand,
-    );
-
     final targets = stringsArg('target');
     if (targets.isEmpty) {
       throwToolExit('No target provided with --target');
@@ -133,6 +129,11 @@ class DevelopCommand extends PatrolCommand {
       flutterCommand: flutterCommand,
     );
     final device = devices.single;
+
+    await _compatibilityChecker.checkVersionsCompatibility(
+      flutterCommand: flutterCommand,
+      targetPlatform: device.targetPlatform,
+    );
 
     // `flutter logs` doesn't work on macOS, so we don't support it for now
     // https://github.com/leancodepl/patrol/issues/1974

--- a/packages/patrol_cli/lib/src/commands/test.dart
+++ b/packages/patrol_cli/lib/src/commands/test.dart
@@ -84,10 +84,6 @@ class TestCommand extends PatrolCommand {
       ),
     );
 
-    await _compatibilityChecker.checkVersionsCompatibility(
-      flutterCommand: flutterCommand,
-    );
-
     final config = _pubspecReader.read();
     final testFileSuffix = config.testFileSuffix;
 
@@ -140,6 +136,11 @@ See https://github.com/leancodepl/patrol/issues/1316 to learn more.
     }
 
     final device = devices.single;
+
+    await _compatibilityChecker.checkVersionsCompatibility(
+      flutterCommand: flutterCommand,
+      targetPlatform: device.targetPlatform,
+    );
 
     final packageName = stringArg('package-name') ?? config.android.packageName;
     final bundleId = stringArg('bundle-id') ?? config.ios.bundleId;

--- a/packages/patrol_cli/lib/src/compatibility_checker.dart
+++ b/packages/patrol_cli/lib/src/compatibility_checker.dart
@@ -1,5 +1,4 @@
 import 'dart:async';
-import 'dart:io' as io;
 
 import 'package:dispose_scope/dispose_scope.dart';
 import 'package:file/file.dart';
@@ -7,6 +6,7 @@ import 'package:patrol_cli/src/base/constants.dart' as constants;
 import 'package:patrol_cli/src/base/exceptions.dart';
 import 'package:patrol_cli/src/base/logger.dart';
 import 'package:patrol_cli/src/base/process.dart';
+import 'package:patrol_cli/src/devices.dart';
 import 'package:patrol_cli/src/runner/flutter_command.dart';
 import 'package:process/process.dart';
 import 'package:version/version.dart';
@@ -28,10 +28,11 @@ class CompatibilityChecker {
 
   Future<void> checkVersionsCompatibility({
     required FlutterCommand flutterCommand,
+    required TargetPlatform targetPlatform,
   }) async {
-    if (io.Platform.isAndroid) {
+    if (targetPlatform == TargetPlatform.android) {
       await _checkJavaVersion(
-        _disposeScope,
+        DisposeScope(),
         _processManager,
         _projectRoot,
         _logger,
@@ -97,24 +98,29 @@ Future<void> _checkJavaVersion(
   Logger logger,
 ) async {
   Version? javaVersion;
-  final javaCompleter = Completer<Version?>();
+  final javaCompleterVersion = Completer<Version?>();
 
   await disposeScope.run((scope) async {
     final process = await processManager.start(
-      ['javac', '--version'],
+      ['flutter', 'doctor', '--verbose'],
       workingDirectory: projectRoot.path,
       runInShell: true,
     )
       ..disposedBy(scope);
 
-    process.listenStdOut((line) async {
-      if (line.startsWith('javac')) {
-        javaCompleter.complete(Version.parse(line.split(' ').last));
-      }
-    }).disposedBy(scope);
+    process.listenStdOut(
+      (line) async {
+        if (line.contains('• Java version') &&
+            javaCompleterVersion.isCompleted == false) {
+          final versionString = line.split(' ').last.replaceAll(')', '');
+          javaCompleterVersion.complete(Version.parse(versionString));
+        }
+      },
+    ).disposedBy(scope);
   });
 
-  javaVersion = await javaCompleter.future;
+  javaVersion = await javaCompleterVersion.future;
+
   if (javaVersion == null) {
     throwToolExit(
       'Failed to read Java version. Make sure you have Java installed and added to PATH',

--- a/packages/patrol_cli/lib/src/compatibility_checker.dart
+++ b/packages/patrol_cli/lib/src/compatibility_checker.dart
@@ -32,6 +32,7 @@ class CompatibilityChecker {
   }) async {
     if (targetPlatform == TargetPlatform.android) {
       await _checkJavaVersion(
+        flutterCommand.executable,
         DisposeScope(),
         _processManager,
         _projectRoot,
@@ -92,6 +93,7 @@ class CompatibilityChecker {
 }
 
 Future<void> _checkJavaVersion(
+  String flutterExecutable,
   DisposeScope disposeScope,
   ProcessManager processManager,
   Directory projectRoot,
@@ -102,7 +104,7 @@ Future<void> _checkJavaVersion(
 
   await disposeScope.run((scope) async {
     final process = await processManager.start(
-      ['flutter', 'doctor', '--verbose'],
+      [flutterExecutable, 'doctor', '--verbose'],
       workingDirectory: projectRoot.path,
       runInShell: true,
     )

--- a/packages/patrol_cli/pubspec.yaml
+++ b/packages/patrol_cli/pubspec.yaml
@@ -1,7 +1,7 @@
 name: patrol_cli
 description: >
   Command-line tool for Patrol, a powerful Flutter-native UI testing framework.
-version: 2.9.0 # Must be kept in sync with constants.dart
+version: 3.0.0  # Must be kept in sync with constants.dart
 homepage: https://patrol.leancode.co
 repository: https://github.com/leancodepl/patrol/tree/master/packages/patrol_cli
 issue_tracker: https://github.com/leancodepl/patrol/issues?q=is%3Aopen+is%3Aissue+label%3A%22package%3A+patrol_cli%22

--- a/packages/patrol_cli/pubspec.yaml
+++ b/packages/patrol_cli/pubspec.yaml
@@ -1,7 +1,7 @@
 name: patrol_cli
 description: >
   Command-line tool for Patrol, a powerful Flutter-native UI testing framework.
-version: 2.8.1 # Must be kept in sync with constants.dart
+version: 2.9.0 # Must be kept in sync with constants.dart
 homepage: https://patrol.leancode.co
 repository: https://github.com/leancodepl/patrol/tree/master/packages/patrol_cli
 issue_tracker: https://github.com/leancodepl/patrol/issues?q=is%3Aopen+is%3Aissue+label%3A%22package%3A+patrol_cli%22

